### ISSUE TITLE
[Linear Layer] Possibly wrong gradient?

### ISSUE THREAD

**Issue by CDLuminate** [torch/nn#1269]
I thought the formulations for the Linear layer should be:

```
column y, Matrix M, column x, column b

y = Wx + b
δ = ∂Loss / ∂y
∂Loss / ∂x = ∂y^T/∂x   ∂Loss/∂y = W^T \cdot δ
∂Loss / ∂b = ∂y^T/∂b   ∂Loss/∂y = eye \cdot δ = δ
∂Loss / ∂W = δ \cdot x^T
```

So I tried to verify those equations in pytorch:
```python
import q
import torch as th
from torch.autograd import Variable as V

a = V(th.rand(2,5), requires_grad=True)
print('a', a)
l = th.nn.Linear(5,3)
y = l.forward(a)
print('y', y)

y_ = th.mm(l.weight, a.t()).t() + l.bias.expand(2,3)
print('y_', y_)
print('||y_-y||_2', (y_-y).norm())

gy = th.rand(2,3)
l.zero_grad()
y.backward(gy)
ga = a.grad
print('ga', ga)
gb = l.bias.grad
print('gb', gb)
gw = l.weight.grad
print('gw', gw)

ga_ = th.mm(l.weight.data.t(), gy.t()).t()
print('ga_', ga_)
gb_ = gy.t().sum(1).t()
print('gb_', gb_)
gw_ = th.mm(gy.t(), a.data)
print('gw_', gw_)

p = lambda x,y: print((x-y).norm())
p(ga.data, ga_)
p(gw.data, gw_)
p(gb.data, gb_)

#q.d()
```

The pytorch version works, and the final three norms are all zero.

However, I tried to verify it in lua torch and it gave me a different result:

```lua
require 'torch'
require 'nn'

x = torch.rand(1, 10)
l = nn.Linear(10, 3) -- insize, outsize
w = l.weight
b = l.bias
gy = torch.ones(1, 3)

print(w:size())
print(b:size())

y = l:forward(x)
l:zeroGradParameters()
gx = l:backward(x, gy)
l:accGradParameters(x, gy)
gb = l.gradBias
gw = l.gradWeight

yy = x * w:t() + b
ggx = gy * w
ggb = gy:t()
ggw = (x:t() * gy):t()

print('yy')
print(yy)
print(y)

print('gx')
print(gx)
print(ggx)

print('gb')
print(gb)
print(ggb)

print('gw')
print(gw)
print(ggw)

-- FIXME: the results differ?
```

The result of the above lua torch code shows, the manually computed weight gradient `ggw`
and the manually computed bias gradient `ggb` are 0.5 times as large as the original `gw` .

```
gb	
 2
 2
 2
[torch.DoubleTensor of size 3]

 1
 1
 1
[torch.DoubleTensor of size 3x1]

gw	
 0.7385  0.0724  0.1870  0.0149  0.7697  1.9020  0.8468  0.6302  0.1299  1.5618
 0.7385  0.0724  0.1870  0.0149  0.7697  1.9020  0.8468  0.6302  0.1299  1.5618
 0.7385  0.0724  0.1870  0.0149  0.7697  1.9020  0.8468  0.6302  0.1299  1.5618
[torch.DoubleTensor of size 3x10]

 0.3693  0.0362  0.0935  0.0075  0.3849  0.9510  0.4234  0.3151  0.0650  0.7809
 0.3693  0.0362  0.0935  0.0075  0.3849  0.9510  0.4234  0.3151  0.0650  0.7809
 0.3693  0.0362  0.0935  0.0075  0.3849  0.9510  0.4234  0.3151  0.0650  0.7809
[torch.DoubleTensor of size 3x10]
```

Is this expected?

**Comment by iamalbert** [comment]
Because you compute and accumulate gradients twice. 
A `backward` call is  `updateGradInput(input, gradOutput)` + `accGradParameters(input,gradOutput,scale)`

**Comment by CDLuminate** [comment]
@iamalbert Thanks, I got it :-)
